// @sentry/utils/esm/memo.js
// tslint:disable:no-unsafe-any
/**
 * Memo class used for decycle json objects. Uses WeakSet if available otherwise array.
 */
export class Memo {
    constructor() {
        // tslint:disable-next-line
        this.hasWeakSet = typeof WeakSet === 'function';
        this.inner = this.hasWeakSet ? new WeakSet() : [];
    }
    /**
     * Sets obj to remember.
     * @param obj Object to remember
     */
    memoize(obj) {
        if (this.hasWeakSet) {
            if (this.inner.has(obj)) {
                return true;
            }
            this.inner.add(obj);
            return false;
        }
        else {
            for (const value of this.inner) {
                if (value === obj) {
                    return true;
                }
            }
            this.inner.push(obj);
            return false;
        }
    }
    /**
     * Removes object from internal storage.
     * @param obj Object to forget
     */
    unmemoize(obj) {
        if (this.hasWeakSet) {
            this.inner.delete(obj);
        }
        else {
            for (let i = 0; i < this.inner.length; i++) {
                if (this.inner[i] === obj) {
                    this.inner.splice(i, 1);
                    break;
                }
            }
        }
    }
}
